drivers: adc: stm32: fix int paths setup

ycsin opened the pull request:
1. Only setup the internal path right before the adc sampling
   1.  If VBAT internal path is setup from `adc_stm32_channel_setup` but never read, then it will be draining current for no reason
   2. To prepare for #47691 which will teardown the internal path after sampling
3. Remove setting up internal path during driver init. The internal path will be setup before sampling if necessary

pending #47758

ycsin committed:
drivers: adc: stm32: Move adc_stm32_setup_channels to start_read

Move the `adc_stm32_setup_channels` function to `start_read`
so that we can guarantee that the internel path is connecte
before reading.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin committed:
drivers: adc: stm32: remove unnecessary VREFINT path

Remove unnecessary VREFINT path connection during the init.
Internal paths should be setup before reading the channel.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

aurel32 reviewed (dismissed):
This looks good to me, tested on STM32WL. And thanks for the second cleanup commit.

aurel32 commented:
My only point is to now if we want to merge the 3 PRs one by one, or if we want to group them in a single PR. In any case it's good to leave the commits separated.

ycsin commented:
I guess yours should be merged first, this pr might cause conflict to everything else

str4t0m commented:
Do I assume correctly that, after the sleeps for channel stabilization have been added in #47758, you'd add those here as well?

ycsin commented:
> Do I assume correctly that, after the sleeps for channel stabilization have been added in #47758, you'd add those here as well?

Yup, I will have to resolve the conflicts

ycsin committed:
drivers: adc: stm32: Use k_usleep variant

Convert these k_sleep calls to use the k_usleep instead.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ycsin commented:
@aurel32 I have rebased the driver and also changed the `k_sleep(K_USEC(X))` to simply `k_usleep(X)`

aurel32 reviewed (dismissed):
This looks good to me. My only concern is that the common path change is done with the ADC enabled, while in PR#47691 you mentioned that the ADC has to be disabled first. In practice it just seems to work (at least on STM32WL)

ycsin commented:
> ADC has to be disabled first

You are right, can we get some insight from ST side? @FRASTM @erwango @ABOSTM

erwango commented:
> > ADC has to be disabled first
> 
> You are right, can we get some insight from ST side? @FRASTM @erwango @ABOSTM
> 
> Probably I should add another commit to disable the ADC so that it will be consistent with the #47691 later

I won't be able to answer on the theory, but I've just performed a non regression from this PR on our test farm and it seems ok.

ycsin committed:
drivers: adc: stm32: Disable ADC before setting common path

According to the reference manual, the ADC should not be
converting when setting the common path, we disable the adc
directly in this driver for the sake of simplicity.

Signed-off-by: Yong Cong Sin <yongcong.sin@gmail.com>

ABOSTM reviewed (approved):
LGTM
